jerry-ext/args/args.c

```diff
@@ -0,0 +1,97 @@
+/* Copyright JS Foundation and other contributors, http://js.foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "jerryscript-ext/args.h"
+#include "jerryscript.h"
+
+/**
+ * It validate the JS arguments and assign them to the native argument.
+ *
+ * @return jerry undefined: the validator passes,
+ *         jerry error: the validator fails.
+ */
+jerry_value_t
+jerryx_validate_and_assign_args (const jerry_value_t *js_arg_p, /**< points to the js argument */
+                                 const jerry_length_t js_arg_cnt, /**< the count of the js arguments */
+                                 const jerryx_arg_t *c_arg_p, /**< points to the jerryx_arg_t */
+                                 const jerry_length_t c_arg_cnt) /**< the count of the native argument */
+{
+  jerry_value_t ret = jerry_create_undefined ();
+
+  if (c_arg_cnt == 0)
+  {
+    return ret;
+  }
+
+  jerryx_arg_iterator_t iterator =
```

Comment:
@martijnthe Please, take a look at the code below. I think it deals with the "challenging use case". And I don't think that you can solve it nicer with iterators either.

```c
typedef struct {
  jerry_value_t callback;
  // .. other native fields
} my_native_obj_t;

static jerry_object_native_info_t my_native_obj_info = {
  // ...
};

static jerryx_arg_result_t
my_native_obj_cb_transform (const jerry_value_t *js_arg_p,
                            jerry_length_t js_arg_cnt,
                            const jerryx_arg_t *c_arg_p)
{
  jerry_value_t js_arg = jerryx_arg_helper_get (js_arg_p, js_arg_cnt, 0);

  if (!jerry_value_is_undefined (js_arg))
  {
    my_native_obj_t *dest = *(my_native_obj_t **) c_arg_p->extra_info;
    jerry_release_value (dest->callback);
    dest->callback = js_arg;
  }

  return jerryx_arg_result_ok (1);
}

static jerry_value_t
my_native_obj_set_cb(const jerry_value_t function_obj,
                     const jerry_value_t this_val,
                     const jerry_value_t args_p[],
                     const jerry_length_t args_count)
{
  my_native_obj_t *native_obj = NULL;

  jerryx_arg_t mapping[] =
  {
    jerryx_arg_native_pointer (&native_obj, &my_native_obj_info, JERRYX_ARG_REQUIRED), // this
    jerryx_arg_custom (NULL, &native_obj, my_native_obj_cb_transform) // arg0: cb function
  };

  jerry_value_t is_ok = jerryx_arg_transform_all (this_val, args_p, args_cnt, mapping, 2);
}
```